frontend/src/main/scala/bloop/dap/DebugSession.scala

```diff
@@ -30,112 +29,119 @@ final class DebugSession(
     ioScheduler: Scheduler
 ) extends DapServer(socket.getInputStream, socket.getOutputStream, DebugExtensions.newContext)
     with Cancelable {
-  type LaunchId = Int
-  private val launches = mutable.Set.empty[LaunchId]
+  private type LaunchId = Int
 
-  private val debugAddress = Promise[InetSocketAddress]()
-  private val exitStatusPromise = Promise[ExitStatus]()
-  private val debuggeeExited = Promise[Unit]()
+  // A set of all processed launched requests by the client
+  private val launchedRequests = mutable.Set.empty[LaunchId]
 
-  private val isStarted = Atomic(false) // if true - the server was already started
-  private val runningDebuggee = Atomic(CancelableFuture.unit)
+  private val debuggeeExited = Promise[Unit]()
+  private val debugAddress = Promise[InetSocketAddress]()
+  private val exitStatusPromise = Promise[DebugSession.ExitStatus]()
 
-  def exitStatus(): Task[ExitStatus] = {
-    Task.fromFuture(exitStatusPromise.future)
-  }
+  private val isStarted = Atomic(false)
+  private val isCancelled = Atomic(false)
 
-  /**
-   * Requests the debuggee to stop. Once that happens, requests the server to disconnect.
-   * When handling the response to this request, the server should close the socket
-   */
-  def cancel(): Unit = {
-    runningDebuggee.get.cancel()
-    Task
-      .fromFuture(debuggeeExited.future)
-      .map(_ => disconnectRequest(InternalRequestId))
-      .foreachL(dispatchRequest)
-      .timeoutTo(FiniteDuration(5, TimeUnit.SECONDS), Task(socket.close()))
-      .runAsync(ioScheduler)
-  }
+  // Access to this cancelable is always protected behind `isCancelled` to avoid race conditions
+  private val runningDebuggee = Atomic(CancelableFuture.unit)
 
   /**
-   * Makes the parent [[run()]] method non-blocking.
-   * Schedules a run-only-once task which:
-   * 1. starts the debuggee
-   * 2. begins listening for debug clients
+   * Schedules the start of the debugging session.
+   *
+   * For a session to start, two executions must happen independently in a
+   * non-blocking way: the debuggee process is started in the background and
+   * the DAP server starts listening to client requests in an IO thread.
    */
-  override def run(): Unit = {
+  def startDebuggeeAndServer(): Unit = {
```

Comment:
If you're worried about people seeing either `run` or `startDebuggeeAndServer` and being confused about which one they should use, we can override `run`, redirect it to `startDebuggeeAndServer` and call the `super.run` inside it.
  
There are some tricks to avoid the invocation of `run` from the outside but they are overkill. The general sentiment driving this change is that we don't want people to be using run but its more precise `startDebuggeeAndServer` because `run` is a huge misnomer about what's happening here :smile:.